### PR TITLE
Reinit typora

### LINKED ISSUE #250200: Package request: typora
**Project description**

Typora is a distraction free markdown editor.
Developed by @abnerlee 

**Metadata**

* homepage URL: https://typora.io/
* source URL: Not open source :( Github Organization: https://github.com/typora
* license: https://support.typora.io/License-Agreement/
* platforms: [linux](https://typora.io/#linux), darwin

### PR DESCRIPTION
**Pull request by npulidomateo**
### Motivation

Typora was removed in pull request #137425 because it refused to run. In issue #250200 it was requested again with a hint that it might run inside an FHS environment. In  #138329 it is said that there is no problem in adding it again if it runs.

closes #250200

### Actual changes 

The derivation of `typora` was re-written from scratch. The `.deb` file is downloaded and unpacked. An FHS environment is created. Typora is run inside the FHS environment.  

I'm adding myself to the maintainers of this package.

*Edit 2023-10-13:* Remove @jensbin from maintainer array.

## Things done

- Built on platform(s)
  - [x] x86_64-linux
  - [ ] aarch64-linux
  - [ ] x86_64-darwin
  - [ ] aarch64-darwin
- [ ] For non-Linux: Is `sandbox = true` set in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
- [ ] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://nixos.org/manual/nixpkgs/unstable/#sec-package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://nixos.org/manual/nixpkgs/unstable/#ssec-nixos-tests-linking) to the relevant packages
- [x] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [x] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [23.11 Release Notes](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2311.section.md) (or backporting [23.05 Release notes](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2305.section.md))
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

Output of `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"` is:
 <details> 
   <summary>1 package built:</summary> 
   <ul> 
    <li>typora</li> 
   </ul> 
 </details>

### CODE REVIEW

**Comment by jensbin** [comment]
Hi @npulidomateo, Thank you. I quickly verified and looks good to me.
Please feel free to remove me from the maintainer array. :)

**Review by jensbin — APPROVED** [review]

**Comment by npulidomateo** [comment]
Hi @jensbin thanks for reviewing! I'll remove you from the maintainer array when this PR is settled.

File: pkgs/applications/editors/typora/default.nix [hunk 1]
```diff
@@ -0,0 +1,103 @@
...
+, alsa-lib
+, buildFHSEnv
+}:
+
+let
+  pname = "typora";
+  version = "1.7.5";
+  src = fetchurl {
+    url = "https://download.typora.io/linux/typora_${version}_amd64.deb";
+    sha256 = "sha256-4Q+fx1kNu98+nxnI/7hLhE6zOdNsaAiAnW6xVd+hZOI=";
```
Comment (wegank) [inline pkgs/applications/editors/typora/default.nix:28]:
```suggestion
    hash = "sha256-4Q+fx1kNu98+nxnI/7hLhE6zOdNsaAiAnW6xVd+hZOI=";
```

File: pkgs/applications/editors/typora/default.nix [hunk 2]
```diff
@@ -0,0 +1,103 @@
...
+    nativeBuildInputs = [ dpkg ];
+
+    dontConfigure = true;
+    dontBuild = true;
+
+    installPhase = ''
+      mkdir -p $out/bin $out/share
+      mv usr/share $out
+      ln -s $out/share/typora/Typora $out/bin/Typora
+    '';
```
Comment (wegank) [inline pkgs/applications/editors/typora/default.nix:44]:
```suggestion
    installPhase = ''
      runHook preInstall
      mkdir -p $out/bin $out/share
      mv usr/share $out
      ln -s $out/share/typora/Typora $out/bin/Typora
      runHook postInstall
    '';
```

File: pkgs/applications/editors/typora/default.nix [hunk 3]
```diff
@@ -0,0 +1,103 @@
...
+let
+  pname = "typora";
+  version = "1.7.5";
+  src = fetchurl {
+    url = "https://download.typora.io/linux/typora_${version}_amd64.deb";
+    sha256 = "sha256-4Q+fx1kNu98+nxnI/7hLhE6zOdNsaAiAnW6xVd+hZOI=";
+  };
+
+  typoraBase = stdenv.mkDerivation {
+    system = "x86_64-linux";
```
Comment (wegank) [inline pkgs/applications/editors/typora/default.nix:32]: What does this line do?

File: pkgs/applications/editors/typora/default.nix (hunk 2 above)
Comment (npulidomateo) [reply pkgs/applications/editors/typora/default.nix]: I forgot those hooks, changed.

File: pkgs/applications/editors/typora/default.nix (hunk 3 above)
Comment (npulidomateo) [reply pkgs/applications/editors/typora/default.nix]: It does nothing, removed.

**Comment by npulidomateo** [comment]
@jensbin I'm removing you from the maintainer array now to avoid further pinging you unnecessarily.

**Comment by v411e** [comment]
@npulidomateo Whooo, this is so cool! Thank you! :tada:

**Pull request merged**